pkg/reconciler/intevents/topic/controller.go

```diff
@@ -85,6 +88,27 @@ func newController(
 		logger.Fatal("Failed to process env var", zap.Error(err))
 	}
 
+	// If there is an error, the projectID will be empty. The reconciler will retry
+	// to get the projectID during reconciliation.
+	projectID, err := utils.ProjectID(os.Getenv(utils.ProjectIDEnvKey), metadataClient.NewDefaultMetadataClient())
+	if err != nil {
+		logging.FromContext(ctx).Error("Failed to get project ID", zap.Error(err))
+	}
+	// Attempt to create a pubsub client for all worker threads to use. If this
+	// fails, pass a nil value to the Reconciler. They will attempt to
+	// create a client on reconcile.
+	client, err := pubsub.NewDefaultPubsubClient(ctx, projectID)
+	if err != nil {
+		logging.FromContext(ctx).Error("Failed to create controller-wide Pub/Sub client", zap.Error(err))
+	}
+
+	if client != nil {
+		go func() {
+			<-ctx.Done()
+			client.Close()
+		}()
+	}
+
```

Comment:
this will not apply to sources as we can specify the project in the CO spec.

Reply:
the idea of having it in the spec was to support cross-project stuff.

Reply:
either way, if you are doing this in the broker, if you cannot create a client here, might be better to fail?

Reply:
After our conversation yesterday, I actually removed this too. See the latest code.

And yes, you are right about the failing the broker. There are some legacy code, I have discussed this with others and opened an issue to revisit whether we should fail it.